shared/ssz/hash_cache.go

```diff
@@ -0,0 +1,248 @@
+package ssz
+
+import (
+	"errors"
+	"fmt"
+	"reflect"
+	"time"
+
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/karlseguin/ccache"
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/prometheus/client_golang/prometheus/promauto"
+	"github.com/prysmaticlabs/prysm/shared/bytesutil"
+	"github.com/prysmaticlabs/prysm/shared/hashutil"
+	"github.com/prysmaticlabs/prysm/shared/params"
+)
+
+var (
+	// ErrNotMerkleRoot will be returned when a cache object is not a merkle root.
+	ErrNotMerkleRoot = errors.New("object is not a merkle root")
+	// maxCacheSize is 2x of the follow distance for additional cache padding.
+	// Requests should be only accessing blocks within recent blocks within the
+	// Eth1FollowDistance.
+	maxCacheSize = params.BeaconConfig().HashCacheSize
+	// Metrics
+	hashCacheMiss = promauto.NewCounter(prometheus.CounterOpts{
+		Name: "powchain_hash_cache_miss",
+		Help: "The number of hash requests that aren't present in the cache.",
+	})
+	hashCacheHit = promauto.NewCounter(prometheus.CounterOpts{
+		Name: "powchain_hash_cache_hit",
+		Help: "The number of hash requests that are present in the cache.",
+	})
+	hashCacheSize = promauto.NewGauge(prometheus.GaugeOpts{
+		Name: "powchain_hash_cache_size",
```

Comment:
why powchain?

Reply:
you are right. no connection to powchain